### ISSUE TITLE
Please make it easier to use PDM with editable dependencies

### ISSUE THREAD

**Issue by whitequark** [pdm-project/pdm#2199]
## Is your feature request related to a problem? Please describe.

At the moment it is not possible to add an editable dependency to `project.dependencies`:

    WARNING: Skipping editable dependency -e chipflow-lib @ file:///${PROJECT_ROOT}/../chipflow-lib in the [project] table. Please move it to the [tool.pdm.dev-dependencies] table

When developing an application using on a library that has no tests or tests with severely insufficient coverage (an unfortunately common occurrence), or just prototyping something, I rely on using the editable version of the library as a dependency. This is a core part of my workflow because it avoids the manual work associated with continually reinstalling the packages, something that is error-prone (you can forget or miss it) and tedious (pdm is pretty fast but it's not instant).

Because the PDM-generated virtualenvs do not include `pip` by default, it is not possible (that I know) to use `pip` to install the dependency just into the virtualenv without ever touching `pyproject.toml` if you weren't prepared for it by passing `--with-pip` when first initializing the check-out, and it would require deleting and recreating the venv, which can take quite a bit of time.

Modifying `pyproject.toml` is fine, but moving the dependency between sections adds noise to the diffs that I look at when I want to look over the state of the working set.

It is frustrating to jump through these hoops each time I want to plug in an editable version of a dependency.

## Describe the solution you'd like

I would like editable dependencies to not be skipped in `project.dependencies`. It is fine if I get a warning each time anyway.

I would like it even more if there was an easy way to substitute an existing dependency for an editable one at a path of my own choosing, without ever touching `pyproject.toml`.

I am happy with any other solution that resolves friction in my workflow.

**Comment by pawamoy** [comment]
I think the proposed workflow here is to add the regular dependency in `project.dependencies` and the editable one in `tool.pdm.dev-dependencies`. When working on your project it will use the editable one, and when building and releasing packages they'll use the regular one.

**Comment by whitequark** [comment]
Oh, you can use *both*? I didn't realize that. I'll check this out, thank you!

Is this documented anywhere?

**Comment by pawamoy** [comment]
I don't think it's documented.

**Comment by whitequark** [comment]
@frostming Do you think you could document this functionality?

**Comment by pawamoy** [comment]
@whitequark Would you like to send a PR? Documentation improvements are always welcome :smile:

**Comment by whitequark** [comment]
@pawamoy Not particularly. I spend much of my time writing documentation for my own projects ([this one](https://github.com/amaranth-lang/amaranth/pull/880) right now) these days and I hope that others would do the same for theirs.

**Comment by whitequark** [comment]
Thanks!

**Comment by whitequark** [comment]
@pawamoy It doesn't look like it works for me:

```
$ pdm lock -G dev
🔒 Lock failed
Unable to find a resolution for amaranth-boards
because of the following conflicts:
  -e file:///${PROJECT_ROOT}/../amaranth-boards#egg=amaranth-boards (from project)
  amaranth-boards @ git+https://github.com/amaranth-lang/amaranth-boards (from project)
To fix this, you could loosen the dependency version constraints in pyproject.toml. See
https://pdm.fming.dev/latest/usage/dependency/#solve-the-locking-failure for more details.
See /tmp/pdm-lock-z_et7qb0.log for detailed debug log.
[ResolutionImpossible]: Unable to find a resolution
Add '-v' to see the detailed traceback
```

**Comment by whitequark** [comment]
@frostming Could you reopen this issue please? It doesn't appear to be fixed.

**Comment by whitequark** [comment]
I've also tried `[tool.pdm.overrides]` which seems like it could help:

```
[tool.pdm.overrides]
amaranth = "-e file:///${PROJECT_ROOT}/../amaranth"
```

but it does not:

```
Updating the lock file...
🔒 Lock successful
Changes are written to pdm.lock.
See /tmp/pdm-install-resolve-5ryhsgao.log for detailed debug log.
[RequirementError]: Expected end or semicolon (after name and no valid version specifier)
    amaranth==@ file:///${PROJECT_ROOT}/../amaranth
```

**Comment by joaomcarlos** [comment]
Hi @frostming this was closed without a proper solution.

What is the correct way to solve this problem?

**TLDR:** have a local package editable for dev and not editable for deployment (in my case its for monorepo, but shouldnt matter?)

Example:
```
dependencies = [
  "lib-common@ file:///${PROJECT_ROOT}/../../packages/common",
]
[tool.pdm.dev-dependencies]
dev = [
    "-e file:///${PROJECT_ROOT}/../../packages/common",
]
```

**Why?** Because I want to ensure `lib-common` get installed properly **and most importantly**, that its sub-dependencies are installed too. While retaining the `editable` ability, so that any changes are applied instantly without running `pdm sync` for every character change.

**Comment by frostming** [comment]
@joaomcarlos 
`pdm install --no-editable`

**Comment by joaomcarlos** [comment]
> @joaomcarlos `pdm install --no-editable`

Inside a dockerfile, `pdm sync --prod --no-editable -v` would be ideal, rather than `pdm install`, correct?

For the ability to get an early error out if theres a mismatch. Or is there a nuance I am not aware of?
